docs/rfc/rfc-007-don't-panic.md

```diff
@@ -0,0 +1,84 @@
+# RFC 007: Don't Panic
+
+## Changelog
+
+- 2021-12-17: initial draft (@tychoish)
+
+## Abstract
+
+Today, the Tendermint core codebase has panics in a number of cases as
+a response to exceptional situations. These panics complicate testing,
+and might make tendermint components difficult to use as a library in
+some circumstances. This document outlines a project of converting
+panics to errors and describes the situations where its safe to
+panic. 
+
+## Background
+
+Panics in Go are a great mechanism for aborting the current execution
+for truly exceptional situations (e.g. memory errors, data corruption,
+processes initialization); however, because they resemble exceptions
+in other languages, it can be easy to over use them in the
+implementation of software architectures. This certainly happened in
+the history of Tendermint, and as we embark on the project of
+stabilizing the package, we find ourselves in the right moment to
+reexamine our use of panics, and largely where panics happen in the
+code base.
+
+There are still some situations where panics are acceptable and
+desireable, but it's important that Tendermint, as a project, comes to
+consensus--perhaps in the text of this document--on the situations
+where it is acceptable to panic. 
+
+### References
+
+- [Defer Panic and Recover](https://go.dev/blog/defer-panic-and-recover)
+- [Why Go gets exceptions right](https://dave.cheney.net/tag/panic)
+- [Don't panic](https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_dont_panic)
+
+## Discussion
+
+### Acceptable Panics 
+
+#### Initialization 
+
+It is unambiguously safe (and desireable) to panic in `init()`
+functions in response to any kind of error. These errors are caught by
+tests, and occur early enough in process initialization that they
+won't cause unexpected runtime crashes. 
+
+Other code that is called early in process initialization MAY panic,
+in some situations if it's not possible to return an error or cause
+the process to abort early, although these situations should be
+vanishingly slim.
+
+#### Data Corruption 
+
+If Tendermint code encounters an inconsistency that could be
+attributed to data corruption or a logical impossibility it is safer
+to panic and crash the process than continue to attempt to make
+progress in these situations. 
+
+Examples including reading data out of the storage engine that
+is invalid or corrupt, or encountering an ambiguous situation where
+the process should halt.
+
+TODO: Add more and better examples here.
+
+#### Request Handling
+
+Code that handles responses to incoming/external requests
```

Comment:
Another case I think is OK, similar to this one, is for packages that generate and recover their own "controlled" panics.  It's fairly common in deeply recursive parsing, query processing, etc., to use a panic to simplify unwinding a "globally-failed" request to the top level.

The user of the package still gets a meaningful error out at the top level, but the implementation avoids the cost of a lot of low-probability bounds checks up and down the stack.

The main distinction between a controlled panic and a more general "request cleanup" handler is that controlled panic handlers should "filter" their recovery: E.g., panic with a specific value type that the handler can switch on. Values that don't match can be propagated along so the system doesn't lose valuable debugging signal. The key properties of controlled panics are that they're (a) intentionally generated, and (b) handled explicitly. The "bad cases" are (a) without (b).

I don't think we have any of this shape in Tendermint right now, but I can think of a couple places we might want to use it:

- **ABCI application failure**: The Application interface does not have any recoverable errors; if the application or the host cannot proceed, the application is expected to panic. We ought to filter and handling these panics cleanly in the host.

- **Complex initialization**: Node initialization has a lot of stages that can fail, but which are notionally "succeed or crash" operations. Using panic rather than aborting the process allows deferred cleanup (so a failure doesn't leave, say, databases in a corrupt state) and can improve legibility.

Reply:
Arguably ABCI calls should allow applications to return errors, though in many cases ABCI errors might want to halt out of safety concerns (and it's a touch unclear, and also when abci apps run in process, their panic rigor might be different from our own, so it's a touch hard to say.

Reply:
Our initialization framework handles errors well (enough, I guess) so I feel decent about not needing to carve out that exception even if it'd be fine.